Find Function Detours

LiddulBOFH opened the pull request:
Added detours for:
FindInBox
FindInSphere
FindInCone (but technically correct)

Also added:
InfMap.ezcoord (turns a chunk coordinate into a string for easy use with entlist table)
InfMap.cleanup_track (cleans old entry in chunk table)
InfMap.update_track (calls above, and then actually writes the new entry)
FindInChunk (returns all of the entities in a chunk)
FindInChunkPostCoord (to be used if you have, for example, indexed using ezcoord

Removed:
Other FindInSphere override

In doing this, there is also a cheap entity tracking system that updates whenever an entity crosses a chunk boundary; old entries are deleted, and if a chunk's table ends up being empty it is removed from the global list
Obviously this only works if an entity actually gets affected by chunk updates

In theory this is cheaper in some cases than the vanilla find functions by using chunks as a way to divide and organize things, HOWEVER these functions do not have limits and can still lag the crap out of the game, so further limitations may be required

Some effort was made to make sure clientside is able to call it, just as you can in vanilla, but salt should still be ingested when it comes to trusting clientside to catch updates

LiddulBOFH commented:
Forgot to add, partially solves #19

Mee12345 commented:
holy shit dude, thanks a ton

only 2 things I would like to change:
1. For the "hardfilter" in the detours use my "InfMap.disable_pickup" table, its basically the same thing but can be overwritten by addon creators (and If spacebuild adopts this base they will likely make their own infinite maps. don't worry about spacebuild not being able to "see" planets, that would be their choice)
2. it would be awesome if you used numbered indexes for vectors (vec[1] instead of vec.x), its a *little* faster because of numerical index instead of string and I want all the optimization possible (attached image has tested "proof")
![image](https://user-images.githubusercontent.com/73554599/204690202-e626dc1a-e5d7-4320-b3b2-f27cf9dfc211.png)
3. I also have a concern.... if someone were to use say.. e2, and called ents.FindInBox(vec(-math.huge), vec(math.huge)), would it crash the game...?

DarianLStephens commented:
Maybe there could be a cvar for the maximum range you can scan, with some reasonable default?

LiddulBOFH commented:
oh, never got a notification for this, yeah I can work on those changes real quick

LiddulBOFH commented:
There, changed filter to use said filter, changed vec indexes (I keep forgetting about that), and as for the distance issue, when I was testing it far earlier I was able to make my game have short hiccups using values of 500,000u, while 100,000 and 200,000u were feasible; I'd want to look into indexing by megachunks as well, and experiment different ways of building tables of chunks to iterate through
I originally was going to go around FindInCone differently, and the way I was doing it would have rapidly been far more expensive (checking corners of each chunk to see if it was visible from origin + angle

LiddulBOFH commented:
So, when it comes to the chunk checking, I eliminate empty chunks immediately to only have a list of chunks that have entries

The vanilla FindInSphere and FindInCone are both built directly on top of FindInBox (FindInSphere at an engine level, and FindInCone I'm not sure)
FindInBox in vanilla iterates over ents.GetAll() which when you consider engine-level entities and whatnot collects a lot more overhead we don't need

My FindInBox function will take a large selection of chunks, determine which ones are outliers and not (after of course eliminating empties), and directly add "internal" chunk entities to the results table, then determine any outlier chunk entities to add to that search or not
I then build on top with FindinSphere to cull by radius (I do want to break it off of FindInBox and cull chunks by radius while determining internal chunks and whatnot, same as FindInBox)
And then I further build on top of FindInSphere with FindInCone for a technically correct conical search, which, in vanilla, uses FindInBox directly and can give results not actually within the given radius

So yes, for a paltry 5-10 entities in play this will have a little more overhead than the vanilla functions, but when there are hundreds of entities (looking at the hi-res prop-by-prop planes and cars and tanks, whatever else fits that bill) I think this will be cheaper

I'm really trying to think big with this, I would love to see this addon take off and see those massive battles happen, search and destroy, whatever else

If you'd like, I'll make a vanilla-esque FindInBox/Sphere/Cone function detour and run both of them in tests with vehicles in various chunks

LiddulBOFH commented:
Okay, ran tests with multiple vehicles across multiple chunks, ents.GetAll() initially is faster until you reach more and more entities at which point culling empty space is faster, however the chunks, searching at the size that I am right now, does get cumbersome at larger expanses; I toyed with the idea last night of further expanding the chunk indexing system I did to work on scales of magnitude, which if handled correctly would further improve performance; I'll experiment with that at some point, but for now I guess I'll move FindInBox to ents.GetAll(), as nasty as it feels

Mee12345 commented:
ah gotcha
ill gladly accept the PR in its current state if that's ok with you (but I will probably trim a bunch of stuff out including the old funcs & "track" system since it causes un-needed overhead)

LiddulBOFH commented:
There is a new find function that'll return everything in a chunk which uses the chunk tracking (which again is only chunk-update based) and doesn't rely on any find functions, simply gives that entry in the table, but the clientside part of the tracking can safely go (meant to turn that back off, will do real quick)

LiddulBOFH commented:
Here, I'm going to close and reopen with changes on an up-to-date branch, it was getting too messy for my taste.
